{% extends 'base.html' %}

{% block static %}
<link rel="stylesheet" href="https://cdn.datatables.net/searchbuilder/1.1.0/css/searchBuilder.dataTables.min.css" />
<link rel="stylesheet" href="https://cdn.datatables.net/buttons/1.7.1/css/buttons.dataTables.min.css" />
<script src="https://cdn.datatables.net/searchbuilder/1.1.0/js/dataTables.searchBuilder.min.js"></script>
<script src="https://cdn.datatables.net/buttons/1.7.1/js/dataTables.buttons.min.js"></script>
{% endblock %}

{% block content %}
	{% include 'includes/sidebar.j2' %}
	<div class="w3-main" style="margin-left:350px">
		<div class="w3-bar w3-flat-silver w3-left-align w3-padding">
			<h3><i class="fas fa-user-tag"></i> <i>{{ sample.samplename }}</i></h3>
		</div>
		<div class="w3-padding">
			<button class="w3-button w3-flat-green-sea w3-hover-flat-turquoise w3-ripple" onclick="save_filters()">Save filters</button>
			<table id="variants" class="w3-table-all" style="width:100%">
				<thead>
					<tr class="w3-flat-green-sea">
						<th class='showTitle control-size-100'>Gene</th>
						<th class='showTitle control-size-100'>Feature</th>
						<th class='showTitle control-size-75'>Can.</th>
						<th class='showTitle control-size-150'>Exon/Intron</th>
						<th class='showTitle control-size-250'>HGVSg</th>
						<th class='showTitle control-size-250'>HGVSc</th>
						<th class='showTitle control-size-250'>HGVSp</th>
						<th class='showTitle control-size-100'>GnomAD</th>
						<th class='showTitle control-size-100'>In Seal (pct)</th>
						<th class='showTitle control-size-100'>In Seal (count)</th>
						<th class='showTitle control-size-300'>ClinSig</th>
						<th class='showTitle control-size-100'>Impact</th>
						<th class='showTitle control-size-300'>Consequence</th>
						<th class='showTitle control-size-100'>MES ref</th>
						<th class='showTitle control-size-100'>MES diff</th>
						<th class='showTitle control-size-100'>MES alt</th>
						<th class='showTitle control-size-100'>Missense</th>
					</tr>
				</thead>
				<tfoot>
					<tr class="w3-flat-green-sea">
						<th class='showTitle control-size-100'>
							Gene
							<span onclick="openmodal(title='Gene', content='The gene symbol (e.g. HGNC) (where available) to the output.', footer='For more information see : <a  href=\'https://www.genenames.org/\' target=\'_blank\' style=\'text-decoration:underline\'><i>HGNC</i></a>')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-100'>
							Feature
							<span onclick="openmodal(title='Feature', content='The feature ID, mainly, corresponding to the RefSeq transcript, or the ensembl ID.', footer='For more information see : <a  href=\'http://grch37.ensembl.org/info/docs/tools/vep/script/index.html\' target=\'_blank\' style=\'text-decoration:underline\'><i>VEP documentation</i></a>')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-75'>
							Can.
							<span onclick="openmodal(title='Canonical', content='Flag if the Feature corresponding to the canonical transcript.<br />Based on MobiDetails API database.', footer='For more information see : <a  href=\'https://mobidetails.iurc.montp.inserm.fr/MDAPI/\' target=\'_blank\' style=\'text-decoration:underline\'><i>MDAPI</i></a>')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-150'>
							Exon/Intron
							<span onclick="openmodal(title='Exon/Intron', content='The number of the exon or intron wich have the variant.', footer='')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
                        <th class='showTitle control-size-250'>
							HGVSg
							<span onclick="openmodal(title='HGVSg', content='The HGVS genomic sequence name', footer='For more information see : <a  href=\'https://varnomen.hgvs.org/\' target=\'_blank\' style=\'text-decoration:underline\'><i>varnomen</i></a>')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
                        <th class='showTitle control-size-250'>
							HGVSc
							<span onclick="openmodal(title='HGVSc', content='The HGVS coding sequence name', footer='For more information see : <a  href=\'https://varnomen.hgvs.org/\' target=\'_blank\' style=\'text-decoration:underline\'><i>varnomen</i></a>')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-250'>
							HGVSp
							<span onclick="openmodal(title='HGVSp', content='The HGVS protein sequence name', footer='For more information see : <a  href=\'https://varnomen.hgvs.org/\' target=\'_blank\' style=\'text-decoration:underline\'><i>varnomen</i></a>')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-100'>
							GnomAD
							<span onclick="openmodal(title='GnomAD', content='The Genome Aggregation Database (gnomAD) is a resource developed by an international coalition of investigators, with the goal of aggregating and harmonizing both exome and genome sequencing data from a wide variety of large-scale sequencing projects, and making summary data available for the wider scientific community.<br /> The dataset used is from release 2.1 (GRCh37/hg19) spans 125,748 exome sequences and 15,708 whole-genome sequences from unrelated individuals sequenced as part of various disease-specific and population genetic studies', footer='For more information see : <a  href=\'https://gnomad.broadinstitute.org\' target=\'_blank\' style=\'text-decoration:underline\'><i>GnomAD website</i></a>')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-100'>
							In Seal (pct)
							<span onclick="openmodal(title='In Seal (pct)', content='The percentage this variant appear in the SEAL database', footer='Warning : this value include this sample')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-100'>
							In Seal (count)
							<span onclick="openmodal(title='In Seal (count)', content='The number of occurences this variant appear in the SEAL database', footer='Warning : this value include this sample')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-300'>
							ClinSig
							<span onclick="openmodal(title='ClinSig', content='ClinVar clinical significance of the dbSNP variant.', footer='For more information see : <a  href=\'https://www.ncbi.nlm.nih.gov/clinvar/docs/clinsig/\' target=\'_blank\' style=\'text-decoration:underline\'><i>ClinSig documentation</i></a>')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-100'>
							Impact
							<span onclick="openmodal(title='Impact', content='IMPACT is calculated by VEP, based on consequences and compatibility with snpEff. <div class=\'w3-flash w3-warning\'><p>Impact categories must be used with care, they were created only to help and simplify the filtering process. Obviously, there is no way to predict whether a HIGH impact or a LOW impact variant is the one producing a phenotype of interest.</p></div>', footer='For more information see : <a  href=\'https://m.ensembl.org/info/genome/variation/prediction/predicted_data.html\' target=\'_blank\' style=\'text-decoration:underline\'><i>VEP documentation</i></a> & <a  href=\'https://pcingola.github.io/SnpEff/se_inputoutput/#impact-prediction\' target=\'_blank\' style=\'text-decoration:underline\'><i>snpEff documentation</i></a>')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-300'>
							Consequence
							<span onclick="openmodal(title='Consequence', content='Consequences are calculated by VEP following SO terms.', footer='For more information see : <a  href=\'https://m.ensembl.org/info/genome/variation/prediction/predicted_data.html\' target=\'_blank\' style=\'text-decoration:underline\'><i>VEP documentation</i></a>')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-100'>
							MES ref
							<span onclick="openmodal(title='MES ref', content='', footer='')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-100'>
							MES diff
							<span onclick="openmodal(title='MES diff', content='', footer='')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-100'>
							MES alt
							<span onclick="openmodal(title='MES alt', content='', footer='')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
						<th class='showTitle control-size-100'>
							Missense
							<span onclick="openmodal(title='Missense', content='Here we calculated the mean from 10 prediction scores. The scores have been selected as the top 10 prediction tools from the dbNSFP v4 publication.<br /><ul><li>CADD</li><li>VEST4</li><li>MetaSVM</li><li>MetaLR</li><li>Eigen</li><li>Eigen-PC</li><li>REVEL</li><li>BayesDel_addAF</li><li>BayesDel_noAF</li><li>ClinPred</li></ul>', footer='For more information see : <a  href=\'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7709417/\' target=\'_blank\' style=\'text-decoration:underline\'><i>dbNSFP v4</i></a>')" class="w3-opacity-max w3-hover-opacity">
								<i class="fas fa-question-circle"></i>
							</span>
						</th>
					</tr>
				</tfoot>
			</table>
		</div>
	</div>

	<!-- MODALS -->
	<div id="help-modal" class="w3-modal">
		<div class="w3-modal-content w3-card-4">
			<header class="w3-container w3-flat-turquoise">
				<span onclick="document.getElementById('help-modal').style.display='none'"
					class="w3-button w3-display-topright">&times;</span>
				<h5 id="help-modal-title"></h5>
			</header>
			<div class="w3-container">
				<p id="help-modal-content"></p>
			</div>
			<footer class="w3-container w3-flat-turquoise">
				<p id="help-modal-footer"></p>
			</footer>
		</div>
	</div>
{% endblock %}


{% block script %}

	const clinsig_dict = {
		"uncertain_significance": {
			"color": "orange",
			"score": 4
		},
		"conflicting_interpretations_of_pathogenicity": {
			"color": "orange",
			"score": 4
		},
		"pathogenic": {
			"color": "alizarin",
			"score": 10
		},
		"likely_pathogenic": {
			"color": "alizarin",
			"score": 7
		},
		"pathogenic/likely_pathogenic": {
			"color": "alizarin",
			"score": 6
		},
		"benign": {
			"color": "turquoise",
			"score": -1.5
		},
		"likely_benign": {
			"color": "turquoise",
			"score": -0.25
		},
		"benign/likely_benign": {
			"color": "turquoise",
			"score": -1
		},
		"not_provided": {
			"color": "peter-river",
			"score": 0
		},
		"other": {
			"color": "peter-river",
			"score": 0
		},
		"protective": {
			"color": "peter-river",
			"score": 0
		},
		"association": {
			"color": "peter-river",
			"score": 0
		},
		"risk_factor": {
			"color": "peter-river",
			"score": 0
		},
		"affects": {
			"color": "peter-river",
			"score": 0
		}
	};

	const impact_dict = {
		"MODERATE": {
			"color": "orange",
			"score" : 2
		},
		"MODIFIER": {
			"color": "orange",
			"score" : 1.5
		},
		"HIGH": {
			"color": "alizarin",
			"score" : 3
		},
		"LOW": {
			"color": "turquoise",
			"score" : 1
		}
	};

	const consequences_dict = {
		"stop_gained": {
			"color": "ff0000",
			"score": 0,
		},
		"stop_lost": {
			"color": "ff0000",
			"score": 0,
		},
		"splice_acceptor_variant": {
			"color": "FF581A",
			"score": 0,
		},
		"splice_donor_variant": {
			"color": "FF581A",
			"score": 0,
		},
		"frameshift_variant": {
			"color": "9400D3",
			"score": 0,
		},
		"transcript_ablation": {
			"color": "ff0000",
			"score": 0,
		},
		"start_lost": {
			"color": "ffd700",
			"score": 0,
		},
		"transcript_amplification": {
			"color": "ff69b4",
			"score": 0,
		},
		"missense_variant": {
			"color": "ffd700",
			"score": 0,
		},
		"protein_altering_variant": {
			"color": "ff0080",
			"score": 0,
		},
		"splice_region_variant": {
			"color": "ff7f50",
			"score": 0,
		},
		"inframe_insertion": {
			"color": "ff69b4",
			"score": 0,
		},
		"inframe_deletion": {
			"color": "ff69b4",
			"score": 0,
		},
		"incomplete_terminal_codon_variant": {
			"color": "ff00ff",
			"score": 0,
		},
		"stop_retained_variant": {
			"color": "76ee00",
			"score": 0,
		},
		"start_retained_variant": {
			"color": "76ee00",
			"score": 0,
		},
		"synonymous_variant": {
			"color": "76ee00",
			"score": 0,
		},
		"coding_sequence_variant": {
			"color": "458b00",
			"score": 0,
		},
		"mature_miRNA_variant": {
			"color": "458b00",
			"score": 0,
		},
		"intron_variant": {
			"color": "02599c",
			"score": 0,
		},
		"NMD_transcript_variant": {
			"color": "ff4500",
			"score": 0,
		},
		"non_coding_transcript_exon_variant": {
			"color": "32cd32",
			"score": 0,
		},
		"non_coding_transcript_variant": {
			"color": "32cd32",
			"score": 0,
		},
		"3_prime_UTR_variant": {
			"color": "7ac5cd",
			"score": 0,
		},
		"5_prime_UTR_variant": {
			"color": "7ac5cd",
			"score": 0,
		},
		"upstream_gene_variant": {
			"color": "a2b5cd",
			"score": 0,
		},
		"downstream_gene_variant": {
			"color": "a2b5cd",
			"score": 0,
		},
		"TFBS_ablation": {
			"color": "a52a2a",
			"score": 0,
		},
		"TFBS_amplification": {
			"color": "a52a2a",
			"score": 0,
		},
		"TF_binding_site_variant": {
			"color": "a52a2a",
			"score": 0,
		},
		"regulatory_region_ablation": {
			"color": "a52a2a",
			"score": 0,
		},
		"regulatory_region_amplification": {
			"color": "a52a2a",
			"score": 0,
		},
		"regulatory_region_variant": {
			"color": "a52a2a",
			"score": 0,
		},
		"feature_elongation": {
			"color": "7f7f7f",
			"score": 0,
		},
		"feature_truncation": {
			"color": "7f7f7f",
			"score": 0,
		},
		"intergenic_variant": {
			"color": "636363",
			"score": 0,
		}
	};

	$(document).ready(function() {
	    table = $('#variants').DataTable({
			buttons:[{
				extend: 'searchBuilder',
				config: {
				    preDefined:{{ filter.filter | safe }},
				}
	        }],
			dom: 'Bfrtip',
			scrollX: true,
    		ajax: '/json/variants/sample/{{ sample.id }}',
    		"columns": [
				{ "className": 'showTitle control-size-100', "data": "annotations.SYMBOL"},
				{ "className": 'showTitle control-size-100', "data": "annotations.Feature"},
				{
					"className": 'showTitle w3-center control-size-75',
					"data": "annotations.canonical",
	                "render": {
						"_": function ( data, type, row ) {
							return data;
						},
						"display": function ( data, type, row ) {
							if (data) {
								response = '<i title="CANONICAL" class="w3-text-flat-peter-river fas fa-star"></i>';
							} else {
								response = '<i title="" class="w3-text-flat-peter-river far fa-star"></i>';
							}
							return response;
		                },
					},
				},
				{
					"className": 'showTitle control-size-150',
					"data": "annotations.EI",
					"render": {
						"_": function ( data, type, row ) {
							return data;
						},
						"sort": function ( data, type, row ) {
							pos_split = data.split(/\s|\//);
							if(pos_split[0] == "Exon") {
								myvar = Number(pos_split[1]).toString().padStart(4, "0") + "_" + Number(pos_split[2]).toString().padStart(4, "0") + "_0";
							} else if (pos_split[0] == "Intron") {
								myvar = Number(pos_split[1]).toString().padStart(4, "0") + "_" + Number(pos_split[2]).toString().padStart(4, "0") + "_1";
							} else {
								myvar = "NA"
							}
							console.log(myvar);
							return myvar
						},
					}
				},
				{ "className": 'showTitle control-size-250', "data": "annotations.HGVSg"},
				{ "className": 'showTitle control-size-250', "data": "annotations.HGVSc"},
				{ "className": 'showTitle control-size-250', "data": "annotations.HGVSp"},
				{
					"className": 'showTitle control-size-100',
					"data": "annotations.GnomADg_max",
					"render" : {
						"_" : function ( data, type, row, meta ) {
							if(data != null) {
					    		return data.toFixed(6);
							} else {
								return null;
							}
					    },
						"display" : function ( data, type, row, meta ) {
							if(data != null) {
					    		return data.toFixed(6);
							} else {
								return "NA";
							}
					    },
	            		"sort" : function ( data, type, row, meta ) {
							if(data != null) {
					    		return data.toFixed(6);
							} else {
								num = -1
								return num.toFixed(6);
							}
					    }
					},
				},
				{
					"className": 'showTitle control-size-125',
					"data": "inseal",
					"render" : {
						"_" : function ( data, type, row, meta ) {
							percent = (data["occurrences"]/data["total_samples"]).toFixed(4);
							return percent;
					    },
					},
				},
				{
					"className": 'showTitle control-size-125',
					"data": "inseal",
					"render" : {
						"_" : function ( data, type, row, meta ) {
							return data["occurrences"];
					    },
						"sort" : function ( data, type, row, meta ) {
							return data["occurrences"];
					    }
					},
				},
				{
					"className": 'showTitle size-300',
					"data": "annotations.CLIN_SIG",
					"render": {
						"_": function ( data, type, row, meta ) {
							cell = {
								"orange": "",
								"alizarin": "",
								"turquoise": "",
								"peter-river": "",
								"asbestos": ""
							}
							style = "style='max-width:200px; overflow:hidden; text-overflow:ellipsis; white-space:nowrap;'";

							for (idx in data) {
								color = data[idx] in clinsig_dict ? clinsig_dict[data[idx]]["color"] : "asbestos"
								cell[color] += " <span class='w3-tag w3-flat-" + color + "' " + style + ">" + data[idx] + "</span>"
							}
							return cell["orange"] + cell["alizarin"] + cell["turquoise"] + cell["peter-river"] + cell["asbestos"];
					    },
						"filter": function ( data, type, row, meta ) {
							return data.toString();
					    },
						"sort": function ( data, type, row, meta ) {
							score = 0;
							for (idx in data) {
								score += data[idx] in clinsig_dict ? clinsig_dict[data[idx]]["score"] : 0
							}
							return score;
					    },
					}
				},
				{
					"className": 'showTitle size-100',
					"data": "annotations.IMPACT",
					"render": {
						"_": function ( data, type, row, meta ) {
							style = "style='max-width:200px; overflow:hidden; text-overflow:ellipsis; white-space:nowrap;'";
							cell = "<span class='w3-tag w3-flat-" + impact_dict[data]["color"] + "' " + style + ">" + data + "</span>";
							return cell;
					    },
						"filter": function ( data, type, row, meta ) {
							return data.toString();
					    },
						"sort": function ( data, type, row, meta ) {
							return impact_dict[data]["score"];
					    },
					}

				},
				{
					"className": 'showTitle size-300',
					"data": "annotations.Consequence",
					"render": {
						"_": function ( data, type, row, meta ) {
							cell = "";
							for (idx in consequences_dict) {
								if (data.includes(idx)){
									color = idx in consequences_dict ? consequences_dict[idx]["color"] : "7f8c8d";
									style = "style='max-width:200px; overflow:hidden; text-overflow:ellipsis; white-space:nowrap;background-color:#" + color + ";'";
									cell += " <span class='w3-tag' " + style + ">" + idx + "</span>"
								}
							}
							return cell;
					    },
						"filter": function ( data, type, row, meta ) {
							return data.toString();
					    },
						"sort": function ( data, type, row, meta ) {
							sort = "A";
							for (idx in consequences_dict) {
								sort += data.includes(idx) ? "1" : "0";
							}
							return sort;
					    }
					},
				},
				{
					"className": 'showTitle size-100',
					"data": "annotations.MaxEntScan_ref",
					"render": {
						"_": function ( data, type, row, meta ) {
							if (data == null){
								return -999;
							}
							return parseFloat(data).toFixed(3);
					    },
						"display": function ( data, type, row, meta ) {
							if (data == null){
								return "NA";
							}
							return parseFloat(data).toFixed(3);
					    },
					},
				},
				{
					"className": 'showTitle size-100',
					"data": "annotations.MaxEntScan_diff",
					"render": {
						"_": function ( data, type, row, meta ) {
							if (data == null){
								return -999;
							}
							return parseFloat(data).toFixed(3);
					    },
						"display": function ( data, type, row, meta ) {
							if (data == null){
								return "NA";
							}
							return parseFloat(data).toFixed(3);
					    },
					},
				},
				{
					"className": 'showTitle size-100',
					"data": "annotations.MaxEntScan_alt",
					"render": {
						"_": function ( data, type, row, meta ) {
							if (data == null){
								return -999;
							}
							return parseFloat(data).toFixed(3);
					    },
						"display": function ( data, type, row, meta ) {
							if (data == null){
								return "NA";
							}
							return parseFloat(data).toFixed(3);
					    },
					},
				},
				{
					"className": 'showTitle size-100',
					"data": "annotations.missenseMean",
					"render": {
						"_" : function ( data, type, row, meta ) {
							if(data != null) {
					    		return data.toFixed(4);
							} else {
								return null;
							}
					    },
						"display" : function ( data, type, row, meta ) {
							if(data != null) {
					    		return data.toFixed(4);
							} else {
								return "NA";
							}
					    },
	            		"sort" : function ( data, type, row, meta ) {
							if(data != null) {
					    		return data.toFixed(4);
							} else {
								num = -1
								return num.toFixed(4);
							}
					    }
					},
				},
			]
		});

	      // Event listener to the two range filtering inputs to redraw on input
	      $('#gnomadMax').keyup( function() {
	          table.draw();
	      } );



	} );

	function save_filters() {
		var d = table.searchBuilder.getDetails();
		console.log(JSON.stringify(d));
		console.log({{ filter.filter | safe }});
	}
	// Script to open and close sidebar
	function w3_open() {
		document.getElementById("sidebar").style.display = "block";
		document.getElementById("overlay").style.display = "block";
	}

	function w3_close() {
		document.getElementById("sidebar").style.display = "none";
		document.getElementById("overlay").style.display = "none";
	}

	$.fn.dataTable.ext.search.push(
	    function( settings, data, dataIndex ) {
	        var gnomadMax = parseFloat( $('#gnomadMax').val());
	        var gnomaAD_AF = parseFloat( data[7] ) || 0; // use data for the age column

	        if ( isNaN( gnomadMax ) || ( gnomaAD_AF <= gnomadMax ) )
	        {
	            return true;
	        }
	        return false;
	    }
	);

	function openmodal(title="Some help", content='Some Content', footer='footer') {
		document.getElementById("help-modal-title").innerHTML = title;
		document.getElementById("help-modal-content").innerHTML = content;
		document.getElementById("help-modal-footer").innerHTML = footer;
		document.getElementById('help-modal').style.display='block'
	}
{% endblock %}
